zerver/views/user_settings.py

```diff
@@ -75,11 +75,18 @@ def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpRes
     assert isinstance(email_change_object, EmailChangeStatus)
     new_email = email_change_object.new_email
     old_email = email_change_object.old_email
+
     with transaction.atomic():
         user_profile = UserProfile.objects.select_for_update().get(
             id=email_change_object.user_profile_id
         )
 
+        if user_profile.delivery_email != old_email:
+            # This is not expected to be possible, since we deactivate
+            # any previous email changes when we create a new one, but
+            # double-check.
+            raise JsonableError(_("Old email address value does not match"))  # nocoverage
```

Comment:
It'd be nice to avoid having translators update a bonus string on account of this theoretical situation. 

I think we can just do `render_confirmation_key_error` with `confirmation/link_expired.html` as the type, since that is what `get_object_from_key` does with revoked confirmations anyway.